AppInspector/Commands/AnalyzeCommand.cs

```diff
@@ -275,467 +255,291 @@ private void ConfigRules()
+                treatEverythingAsCode = _options.TreatEverythingAsCode,
+                confidenceFilter = _confidence
+            };
+
+            _rulesProcessor = new RuleProcessor(rulesSet, rpo);
 
             //create metadata helper to wrap and help populate metadata from scan
-            _metaDataHelper = new MetaDataHelper(_options.SourcePath, !_options.AllowDupTags);
+            _metaDataHelper = new MetaDataHelper(_options.SourcePath, false);
         }
 
         #endregion configureMethods
 
         /// <summary>
-        /// Main entry point to start analysis from CLI; handles setting up rules, directory enumeration
-        /// file type detection and handoff
-        /// Pre: All Configure Methods have been called already and we are ready to SCAN
+        /// Populate the MetaDataHelper with the data from the FileEntries.
         /// </summary>
-        /// <returns></returns>
-        public AnalyzeResult GetResult()
+        /// <param name="cancellationToken"></param>
+        /// <param name="opts"></param>
+        /// <param name="populatedEntries"></param>
+        public AnalyzeResult.ExitCode PopulateRecords(CancellationToken cancellationToken, AnalyzeOptions opts, IEnumerable<FileEntry> populatedEntries)
         {
-            WriteOnce.SafeLog("AnalyzeCommand::Run", LogLevel.Trace);
-            WriteOnce.Operation(MsgHelp.FormatString(MsgHelp.ID.CMD_RUNNING, "Analyze"));
-            AnalyzeResult analyzeResult = new AnalyzeResult()
-            {
-                AppVersion = Utils.GetVersionString()
-            };
+            WriteOnce.SafeLog("AnalyzeCommand::PopulateRecords", LogLevel.Trace);
 
-            try
+            if (_rulesProcessor is null || populatedEntries is null)
             {
-                _metaDataHelper?.Metadata.IncrementTotalFiles(_srcfileList.Count());//updated for zipped files later
+                return AnalyzeResult.ExitCode.CriticalError;
+            }
 
-                //lamda1 for choosing single or multi-threaded processing just runs rules analysis on filename arg
-                Action<string> AnalyzeFiles = filename =>
+            if (opts.SingleThread)
+            {
+                foreach (var entry in populatedEntries)
                 {
-                    if (new FileInfo(filename).Length == 0)
-                    {
-                        WriteOnce.SafeLog(MsgHelp.FormatString(MsgHelp.ID.ANALYZE_EXCLUDED_TYPE_SKIPPED, filename), LogLevel.Warn);
-                        _metaDataHelper?.Metadata.IncrementFilesSkipped();
-                        return;
-                    }
+                    if (cancellationToken.IsCancellationRequested) { break; }
+                    ProcessAndAddToMetadata(entry);
+                }
+            }
+            else
+            {
+                Parallel.ForEach(populatedEntries, new ParallelOptions() { CancellationToken = cancellationToken }, entry => ProcessAndAddToMetadata(entry));
```

Comment:
This is voluntary task completing with a cancellation token. I dont know if it makes sense to do a preemptive one when the max allowed run time (an universal timer when it expires, exit process). What do you think?

Reply:
Fixed not catching the operation cancelled exception here. As discussed should now behave as expected.

Reply:
This will throw here which will be caught and get result will return Canceled as the status.